api/docs/multi_trace_window.dox

```diff
@@ -0,0 +1,396 @@
+
+The worst case inaccuracy here is a thread who starts writing in window N but ends up
+unscheduled until a much later window M.  But at most one basic block's worth of
+trace entries will be placed into window N even though it happened later.  Thus we
+have "basic block accuracy", which is pretty good, as typically a basic block only
+contains a few instructions.
+
+## Proposal A: Separate raw files split at flush time
+
+If we're splitting raw files (see above), we would use the end-of-block window-change
+flush to emit a thread exit and create a new file.  In post-processing, we'd add any
+missing thread exits to threads that don't have them, to cover waiting threads who
+never reached a flush.
+
+As discussed above, the trigger thread would create a new directory for each window.
+A just-finished buffer is writtent to the directory corresponding to the window for
+its start point.
+
+A thread that is unscheduled for a long time could have a nearly-full buffer that is
+not written out until many windows later, but it would be written to the old
+directory for the old window.  The next buffer would go to a new file in the new
+window, with no files in the in-between window directories.
+
+(Originally we thought this scheme would have buffer-level inaccuracy (and talked
+about using timestamps at the start and end of each buffer to detect): but that would
+only be if it wrote out to the current window dir.)
+
+## Proposal B: Label buffers with owning window
+
+If we add the window ordinal to every buffer header, we can identify which window
+they belong to, and avoid the need to separate raw files.  A window-end flush ensures
+a buffer belongs solely to the window identified in its header; the next buffer will
+have the new window value.
+
+This scheme can be used with file splitting during raw2trace, or waiting until
+analysis.  Each thread has one raw file which contains all windows during the
+execution.
+
+## Proposal C: Trigger thread identifies buffer transition point of the other threads
+
+For this proposal, the thread triggering the end of the window walks the other
+threads and identifies the phase transition point inside the buffer, telling the
+post-processor where to split them.
+
+I considered having the triggerer also flush the buffers, but that is challenging
+with a race with the owner also flushing.  Plus, it still requires post-processing
+help to identify the precise point for splitting the buffer (without synchronization
+the triggerer can only get close).
+
+To avoid barriers on common case trace buffer writes, we use a lazy scheme where the
+triggerer does not modify the trace buffers themselves, but instead marks which
+portion has been written using a separate variable never accessed in a fastpath.
+
+Implementation:
+
++ The tracer maintains a global list of thread buffers using a global mutex on thread
+  init and exit.
+
++ Each trace buffer has a corresponding externally_written variable holding a
```

Comment:
We may need multiple such externally written variables if there are multiple window transitions while a thread is still filling up a single buffer.

Reply:
Done.